### PR TITLE
Re-implement processes route to restore compatibility with the current cloud health checks

### PR DESCRIPTION
**Pull request by aleksmaus**
## What does this PR do?

Reimplemented the ```/processes``` routes as minimum sufficient health check for the cloud.

Addresses "V2: 8.6.0-SNAPSHOT not starting in ESS"
https://github.com/elastic/elastic-agent/issues/1731

This doesn't reimplement the full `/stats` collection functionality for example, because there is a new V2 way to check the health and stats that is decoupled from internal child processes handling implementation.
This change is added just to fix the cloud deployments health breakage short term. In the long term cloud would need to be updated for V2, @blakerouse can provide more details.

We had a brief conversation today about this today and reimplementing ```processes``` route fully would take longer, breaks V2 model and might not be necessary. It would be nice if @simitt or @gcpagano could validate the assumptions that this API implementation is sufficient, based on the sample request/response examples below.

Unfortunately I have not had a chance to test this with the actual cloud deployment yet (yet to figure how to) and will be out till the next week. @cmacknz, @blakerouse if I don't get it tested later tonight, could you please take care of this part?

## Why is it important?

Fixed the broken "V2: 8.6.0-SNAPSHOT not starting in ESS"

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/1731

## Examples of requests/responses with the `/processes` route

```
➜  curl -v 127.0.0.1:6791/processes
{"processes":[{"id":"beat/metrics-monitoring","pid":"metricbeat","binary":"80401","source":{"kind":"internal","outputs":["monitoring"]}},{"id":"http/metrics-monitoring","pid":"metricbeat","binary":"80402","source":{"kind":"internal","outputs":["monitoring"]}},{"id":"system/metrics-default","pid":"metricbeat","binary":"80395","source":{"kind":"configured","outputs":["default"]}},{"id":"osquery-default","pid":"osquerybeat","binary":"80396","source":{"kind":"configured","outputs":["default"]}},{"id":"log-default","pid":"filebeat","binary":"80397","source":{"kind":"configured","outputs":["default"]}},{"id":"winlog-default","pid":"filebeat","binary":"80399","source":{"kind":"configured","outputs":["default"]}},{"id":"filestream-monitoring","pid":"filebeat","binary":"80400","source":{"kind":"internal","outputs":["monitoring"]}}]}%

➜  curl -v 127.0.0.1:6791/processes/osquery-default
{"state":"healthy","message":"Healthy: communicating with pid '80396'"}%

➜  curl -v 127.0.0.1:6791/processes/foobar-something
* Connected to 127.0.0.1 (127.0.0.1) port 6791 (#0)
> GET /processes/foobar-something HTTP/1.1

< HTTP/1.1 404 Not Found
< Content-Type: application/json; charset=utf-8
< Date: Tue, 22 Nov 2022 21:39:00 GMT
< Content-Length: 78
<
{"type":"UNEXPECTED","reason":"matching component foobar-something not found"}
```

### CODE REVIEW

**Review by blakerouse — APPROVED** [review]
Looks good to me.

Hopefully it works with cloud for now. Want to remove as soon as possible, once they transition to the health check endpoint.

File: internal/pkg/agent/application/monitoring/server.go [hunk 1]
```diff
@@ -54,6 +59,13 @@ func exposeMetricsEndpoint(
 	statsHandler := statsHandler(ns("stats"))
 	r.Handle("/stats", createHandler(statsHandler))
 
+	if enableProcessStats {
+		r.Handle("/processes", createHandler(processesHandler(coord)))
+		r.Handle("/processes/{processID}", createHandler(processHandler(coord, statsHandler)))
+		r.Handle("/processes/{processID}/", createHandler(processHandler(coord, statsHandler)))
+		r.Handle("/processes/{processID}/{beatsPath}", createHandler(processHandler(coord, statsHandler)))
```
Comment (cmacknz) [inline internal/pkg/agent/application/monitoring/server.go:66]: What is `{beatsPath}` here? I don't see it used in the handler.

**Review by cmacknz — APPROVED** [review]
LGTM, seems to work locally. Thanks!

File: internal/pkg/agent/application/monitoring/server.go (hunk 1 above)
Comment (aleksmaus) [reply internal/pkg/agent/application/monitoring/server.go]: removed

**Comment by gcpagano** [comment]
I think this looks OK from an health check point of view, we call `/processes` and we only check for the status code to be a success.

We call `/processes` also when monitoring APM server mode (when a user migrates to managed mode in Kibana we detect that and do a bunch of stuff on our side).
That seems OK to, we look for an `id:  apm-server` in the response from `/processes` https://github.com/elastic/cloud/blob/0831e29bbb3822139b5078d879e568c141f72e8f/scala-services/runner/src/main/scala/no/found/runner/allocation/stateless/ApmServerModePublisher.scala#L139-L174, but I guess it's worth double-checking it @aleksmaus 

It seems metrics will be broken, as we use endpoints like `/processes/apm-server-es-containerhost/stats` and `/processes/fleet-server-es-containerhost/stats`

**Comment by simitt** [comment]
@aleksmaus @cmacknz what would be the new way of collecting the metrics with v2?

**Comment by cmacknz** [comment]
> It seems metrics will be broken, as we use endpoints like /processes/apm-server-es-containerhost/stats and /processes/fleet-server-es-containerhost/stats

@gcpagano if we merge this as is, will integrations server be usable on cloud? Do we just need to get past the health check to unblock further testing? If that is the case I think we should merge this as is and follow up with fixing the metrics reporting.

> @aleksmaus @cmacknz what would be the new way of collecting the metrics with v2?

In V2 we do not want to know anything about the processes we run, so the ideal way would be to use metricbeat to collect stats either as part of the agent's built in monitoring or separately. This is why the stats endpoint was removed, it requires building knowledge of how to collect stats into the agent itself. 

However I don't see adding the overhead of a monitoring metricbeat into integrations server or forcing cloud to setup a stack monitoring metricbeat on short notice as a great solution here, so probably we need to figure out how to restore the /stats endpoint as well. We didn't realize the importance of the processes endpoint when we removed it.

**Comment by gcpagano** [comment]
> > It seems metrics will be broken, as we use endpoints like /processes/apm-server-es-containerhost/stats and /processes/fleet-server-es-containerhost/stats
> 
> @gcpagano if we merge this as is, will integrations server be usable on cloud? Do we just need to get past the health check to unblock further testing? If that is the case I think we should merge this as is and follow up with fixing the metrics reporting.

I think resorting `processes` to fix the health check should be enough. Metrics won't work but that won't stop integrations server from running.

**Pull request merged**